File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        const result = []


        const dfs = (sum, combination, index) => {
            if (target === sum) {
                return result.push([...combination])
            }
            if (target < sum || index >= nums.length) {
                return
            }
            combination.push(nums[index])
            dfs(sum + nums[index], combination, index)

            combination.pop()
            dfs(sum, combination, index + 1)
        }

        dfs(0, [], 0)

        return result;
    }

}

File: Data Structures & Algorithms/three-integer-sum/submission-0.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        result = []

        j = 0

        while j < len(nums) - 2:
            k = j + 1
            l = len(nums) - 1

            while k < l:
                total = nums[j] + nums[k] + nums[l]

                if total == 0:
                    result.append([nums[j], nums[k], nums[l]])

                    k += 1
                    l -= 1

                    while k < l and nums[k] == nums[k - 1]:
                        k += 1

                    while k < l and nums[l] == nums[l + 1]:
                        l -= 1

                elif total < 0:
                    k += 1

                else:
                    l -= 1

            j += 1
            while j < len(nums) - 2 and nums[j] == nums[j - 1]:
                j += 1

        return result
